File: src/proto/example/module_example.py
#!/usr/bin/python
# add path to the proto messages
import sys
sys.path.append('../../')

# import the proto messages
import proto.python.ast_pb2 as ast_pb
from proto.python.ast_pb2 import AstNode, AstLookupConfig, FileInfo
from proto.python.module_pb2 import ModuleResult, ModuleSummary, ModuleStatic
from google.protobuf.text_format import MessageToString, Merge


def write_proto_to_file(proto, filename, binary=True):
    if binary:
        f = open(filename, "wb")
        f.write(proto.SerializeToString())
        f.close()
    else:
        f = open(filename, "w")
        f.write(MessageToString(proto))
        f.close()


def read_proto_from_file(proto, filename, binary=True):
    if binary:
        f = open(filename, "rb")
        proto.ParseFromString(f.read())
        f.close()
    else:
        f = open(filename, "r")
        Merge(f.read(), proto)
        f.close()


def get_mock_source_node():
    source_node = AstNode()
    # invocation detail
    source_node.type = AstNode.FUNCTION_DECL_REF_EXPR
    source_node.name = "read"
    source_node.full_name = "file.read"
    source_node.base_type = "file"
    source_node.arguments.extend(["\"/home/maloss/.ssh/id_rsa\"", "\"r\""])
    # TODO: look up the id in the config protobuf
    source_node.id = 22
    # source code location
    source_file_info = FileInfo()
    source_file_info.filename = "source_filename.py"
    source_file_info.relpath = "relpath/to/dir"
    source_node.range.start.row = 10
    source_node.range.start.column = 20
    source_node.range.start.file_info.CopyFrom(source_file_info)
    source_node.range.end.row = 10
    source_node.range.end.column = 40
    source_node.range.end.file_info.CopyFrom(source_file_info)
    return source_node


def get_mock_sink_node():
    sink_node = AstNode()
    # invocation detail
    sink_node.type = AstNode.FUNCTION_DECL_REF_EXPR
    sink_node.name = "send"
    sink_node.full_name = "network.send"
    sink_node.base_type = "network"
    sink_node.arguments.extend(["\"https://www.malicious.com\"", "\"non-blocking\""])
    # TODO: look up the id in the config protobuf
    sink_node.id = 88
    # source code location
    sink_file_info = FileInfo()
    sink_file_info.filename = "sink_filename.py"
    sink_file_info.relpath = "relpath/to/dir"
    sink_node.range.start.row = 200
    sink_node.range.start.column = 8
    sink_node.range.start.file_info.CopyFrom(sink_file_info)
    sink_node.range.end.row = 200
    sink_node.range.end.column = 100
    sink_node.range.end.file_info.CopyFrom(sink_file_info)
    return sink_node


def get_mock_propagate_node():
    # TODO: add propagate node if available
    propagate_node = AstNode()
    return propagate_node


def get_mock_danger_node():
    danger_node = AstNode()
    danger_node.type = AstNode.FUNCTION_DECL_REF_EXPR
    danger_node.name = "exec"
    danger_node.full_name = "subprocess.exec"
    danger_node.base_type = "subprocess"
    danger_node.arguments.extend(["\"rm\"", "\"-rf\"", "\"/\""])
    # TODO: look up the id in the config protobuf
    danger_node.id = 66
    # source code location
    danger_file_info = FileInfo()
    danger_file_info.filename = "danger_filename.py"
    danger_file_info.relpath = "relpath/to/dir"
    danger_node.range.start.row = 888
    danger_node.range.start.column = 0
    danger_node.range.start.file_info.CopyFrom(danger_file_info)
    danger_node.range.end.row = 888
    danger_node.range.end.column = 99
    danger_node.range.end.file_info.CopyFrom(danger_file_info)
    return danger_node


def set_result(result):
    # skip package_info for now
    source_node = get_mock_source_node()
    sink_node = get_mock_sink_node()
    propagate_node = get_mock_propagate_node()
    flow = result.flows.add()
    flow.source.CopyFrom(source_node)
    flow.source.source_type = ast_pb.SOURCE_FILE
    flow.sink.CopyFrom(sink_node)
    flow.sink.sink_type = ast_pb.SINK_NETWORK
    flow.hops.add().CopyFrom(propagate_node)

    danger_node = get_mock_danger_node()
    danger = result.dangers.add()
    danger.danger.CopyFrom(danger_node)


def set_summary(summary):
    # skip package_info for now
    # there is no id for new sources
    source = summary.sources.add()
    source.node.type = AstNode.FUNCTION_DECL
    source.node.name = "get"
    source.node.full_name = "requests.get"
    source.node.base_type = "requests"
    source.node.arguments.extend(["url", "mode"])
    # TODO: set the source range for source.node
    reachable_old_source = source.reachable_sources.add()
    reachable_old_source.CopyFrom(get_mock_source_node())
    reachable_old_source.source_type = ast_pb.SOURCE_NETWORK

    # the add() function returns reference
    sink = summary.sinks.add()
    sink.node.type = AstNode.FUNCTION_DECL
    sink.node.name = "post"
    sink.node.full_name = "requests.post"
    sink.node.base_type = "requests"
    sink.node.arguments.extend(["url", "mode", "data"])
    # TODO: set the source range for sink.node
    reachable_old_sink = sink.reachable_sinks.add()
    reachable_old_sink.CopyFrom(get_mock_sink_node())
    reachable_old_sink.sink_type = ast_pb.SINK_NETWORK


# load the astgen config from file
config = AstLookupConfig()
read_proto_from_file(config, '../../../config/astgen_python_smt.config', binary=False)
print("loaded config with %d apis to check!" % len(config.apis))


# initialize result and summary
result = ModuleResult()
summary = ModuleSummary()
static = ModuleStatic()

# compute and fill the results into protobuf
set_result(result)
set_summary(summary)
static.flows.MergeFrom(result.flows)
static.dangers.MergeFrom(result.dangers)
static.sources.MergeFrom(summary.sources)
static.sinks.MergeFrom(summary.sinks)

module_result_pb_fname = "module_result_py.pb"
module_result_txt_fname = "module_result_py.txt"
module_summary_pb_fname = "module_summary_py.pb"
module_summary_txt_fname = "module_summary_py.txt"
module_static_pb_fname = "module_static_py.pb"
module_static_txt_fname = "module_static_py.txt"

# output protobuf message, by default, the output is binary format
write_proto_to_file(result, module_result_pb_fname)
write_proto_to_file(summary, module_summary_pb_fname)
write_proto_to_file(static, module_static_pb_fname)
print("saved result to %s in binary format!" % module_result_pb_fname)
print("saved summary to %s in binary format!" % module_summary_pb_fname)
print("saved static to %s in binary format!" % module_static_pb_fname)

# output protobuf message, use the binary argument to generate text output
write_proto_to_file(result, module_result_txt_fname, binary=False)
write_proto_to_file(summary, module_summary_txt_fname, binary=False)
write_proto_to_file(static, module_static_txt_fname, binary=False)
print("saved result to %s in text format!" % module_result_txt_fname)
print("saved summary to %s in text format!" % module_summary_txt_fname)
print("saved static to %s in text format!" % module_static_txt_fname)

File: src/proto/example/module_example.js
var fs = require("fs");
var ast_pb = require("../javascript/ast_pb");
var module_pb = require("../javascript/module_pb");
var cp = require('child_process'); 

var results;

function getSourceNode(package_results) {
    	var sourceNode = new ast_pb.AstNode();
	
	var source = package_results.sources[0]
	
	sourceNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL_REF_EXPR)
	sourceNode.setName(source.property)
		 
	    sourceNode.setFullName(source.object + "." + source.property)
	    sourceNode.setBaseType(source.object)
	    sourceNode.setArgumentsList(source.argument_list)
	    
    return sourceNode;
}

function getSinkNode(package_results) {
    var sinkNode = new ast_pb.AstNode();

    var sink = package_results.sinks[0] 

    sinkNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL_REF_EXPR)
	    sinkNode.setName(sink.property)
	    sinkNode.setFullName(sink.object + "." + sink.property)
	    sinkNode.setBaseType(sink.object)
	    sinkNode.setArgumentsList(sink.argument_list)

    return sinkNode;
}

function getPropagateNode(package_results) {
    var propagateNode = new ast_pb.AstNode();
    return propagateNode;
}

function getDangerNode() {
    var dangerNode = new ast_pb.AstNode();

	    dangerNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL_REF_EXPR)
	    dangerNode.setName("exec")
	    dangerNode.setFullName("subprocess.exec")
	    dangerNode.setBaseType("subprocess")
	    dangerNode.setArgumentsList(["\"rm\"", "\"-rf\"", "\"/\""])
	    // TODO: look up the id in the config protobuf
	    dangerNode.id = 66
	    // source code location
	    dangerFileInfo = new ast_pb.FileInfo()
	    dangerFileInfo.filename = "danger_filename.py"
	    dangerFileInfo.relpath = "relpath/to/dir"

    return dangerNode;
}

function getSourceSummary(moduleSource, package_results, file_name, file_location) {

    var file_name_stripped = file_name.replace(".js", "")
	
	var sourceNode = new ast_pb.AstNode();
	
	var source = package_results.sources[0]
	
	sourceNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL)
	sourceNode.setName(source.property)
		 
	    sourceNode.setFullName(file_name_stripped + "." + source.property)
	    sourceNode.setBaseType(file_name_stripped)
	    sourceNode.setArgumentsList(source.argument_list)

    // Add reachable nodes
    var new_source = moduleSource.addReachableSources()
     
    var reachable_node = new ast_pb.AstNode();
    
    reachable_node.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL_REF_EXPR)
    reachable_node.setName(source.reachable_property)
    reachable_node.setFullName(source.reachable_object + "." + source.reachable_property)
    reachable_node.setBaseType(source.reachable_object)
    reachable_node.setArgumentsList(source.reachable_argument_list)

    var file_info = new ast_pb.FileInfo
    file_info.setFilename(file_name)
    file_info.setRelpath(file_location)

    var begin_loc = new ast_pb.SourceLocation
    begin_loc.setRow(source.row);
    begin_loc.setColumn(source.startColumn);
    begin_loc.setFileInfo(file_info);

    var end_loc = new ast_pb.SourceLocation
    end_loc.setRow(source.row);
    end_loc.setColumn(source.endColumn);
    end_loc.setFileInfo(file_info);

    var test_range = new ast_pb.SourceRange
    test_range.setStart(begin_loc)
    test_range.setEnd(end_loc)
     
    reachable_node.setRange(test_range)
    
    var id
    for (var i = 0; i < config.array[0].length; i++) {
	node = config.array[0][i]
	if (node[2] == source.object + "." + source.property) {
		id = node[12]
	}
    }
    reachable_node.setId(id)

    new_source.setNode(reachable_node)   
    moduleSource.setNode(sourceNode);
    return moduleSource;
}

function getSinkSummary(moduleSink, package_results, file_name, file_location) {
   
    var file_name_stripped = file_name.replace(".js", "")
    var sinkNode = new ast_pb.AstNode();

    var sink = package_results.sinks[0] 
    sinkNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL)
    sinkNode.setName(sink.property)
    sinkNode.setFullName(file_name_stripped + "." + sink.property)
    sinkNode.setBaseType(file_name_stripped)
    sinkNode.setArgumentsList(sink.argument_list)

    // Add reachable nodes
    var new_sink = moduleSink.addReachableSinks()
     
    var reachable_node = new ast_pb.AstNode();

    reachable_node.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL_REF_EXPR)
    reachable_node.setName(sink.reachable_property)
    reachable_node.setFullName(sink.reachable_object + "." + sink.reachable_property)
    reachable_node.setBaseType(sink.reachable_object)
    reachable_node.setArgumentsList(sink.reachable_argument_list)

    var file_info = new ast_pb.FileInfo
    file_info.setFilename(file_name)
    file_info.setRelpath(file_location)

    var begin_loc = new ast_pb.SourceLocation
    begin_loc.setRow(sink.row);
    begin_loc.setColumn(sink.startColumn);
    begin_loc.setFileInfo(file_info);

    var end_loc = new ast_pb.SourceLocation
    end_loc.setRow(sink.row);
    end_loc.setColumn(sink.endColumn);
    end_loc.setFileInfo(file_info);

    var test_range = new ast_pb.SourceRange
    test_range.setStart(begin_loc)
    test_range.setEnd(end_loc)
     
    //reachable_node.range = test_range
    reachable_node.setRange(test_range)
   
    var id
    for (var i = 0; i < config.array[0].length; i++) {
	    node = config.array[0][i]
	    if (node[2] == sink.object + "." + sink.property) {
		    id = node[12]
	    }
    }
    reachable_node.setId(id)

    new_sink.setNode(reachable_node) 
    moduleSink.setNode(sinkNode);
   
    return moduleSink;
}

function setResult(result, package_results) {
    
    var sourceNode = getSourceNode(package_results);
    var sinkNode = getSinkNode(package_results);
    //var propagateNode = getPropagateNode(package_results);
    sourceNode.setSourceType(ast_pb.SOURCE_FILE);
    sinkNode.setSinkType(ast_pb.SINK_NETWORK);
    var flow = new module_pb.ModuleFlow();
    flow.setSource(sourceNode);
    flow.setSink(sinkNode);
    flow.setSource(flowSource);
    flow.setSink(flowSink);
    result.addFlows(flow);

    /*var dangerNode = getDangerNode();
    var danger = new module_pb.ModuleDanger();
    danger.setSink(dangerNode)
    result.addDangers(danger);*/

}

function setSummary(summary, package_results, file_name, file_location) {
    
    //var sourceNode = new ast_pb.AstNode();
    //sourceNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL);
	
    var moduleSource = new module_pb.ModuleSource();
    moduleSource = getSourceSummary(moduleSource, package_results, file_name, file_location);
         
    //var sinkNode = new ast_pb.AstNode();
    //sinkNode.setType(ast_pb.AstNode.NodeType.FUNCTION_DECL);
    
    var moduleSink = new module_pb.ModuleSink();
    moduleSink = getSinkSummary(moduleSink, package_results, file_name, file_location);	

    summary.addSources(moduleSource);
    summary.addSinks(moduleSink);
}

// load the astgen config from file
var configStr = fs.readFileSync("../../../config/astgen_javascript_smt.config.pb");
var config = ast_pb.AstLookupConfig.deserializeBinary(configStr);
console.log("loaded config with " + config.getApisList().length + " apis to check!");

// Process input
var target_package = process.argv[2]
processInput(target_package, target_package, config.getApisList(), function(package_results, file_name, file_location) {
	processResults(package_results, file_name, file_location)
})


function processInput(input, original_input, source_sink_list, callback) {
	var analyze_result;
	if (fs.lstatSync(input).isDirectory()) {
		fs.readdir(input, function(err, files) {
			if (err) return console.log(err);

			files.forEach( function(file) {
				processInput(input + file, original_input, source_sink_list, callback) //input + "/" + file
			});
		});
	}
	else {
		if (input.substring(input.length - 3, input.length) == '.js') {
			fs.readFile(input, 'utf8', function (err, code) {
				if (err) {
					throw err;
				}

				source_sink_list = JSON.stringify(source_sink_list)
				var ls = cp.fork("../../../../jsprime/engine.js", [code, source_sink_list])
				var timeout_val = setTimeout(function() { ls.kill() }, 20000);
				
				ls.on('message', (msg) => {
					  callback(JSON.parse(msg), input, original_input)
					  clearTimeout(timeout_val)
				});
			});
		}
	}
}


function processResults(package_results, file_location, original_location) {


	file_name = file_location.split('/').pop()
	relative_path = "/" + original_location.replace(file_location, "")
	
	if (package_results.malicious == 'true') {

		// initialize result and summary
		var result = new module_pb.ModuleResult();
		var summary = new module_pb.ModuleSummary();

		// compute and fill the results into protobuf
		setResult(result, package_results);
		setSummary(summary, package_results, file_name, relative_path);

		var package_name = target_package.split('/').pop()

		// output protobuf message in binary format
		var moduleResultPbFname = package_name + "_result_js.pb";
		var moduleSummaryPbFname = package_name + "_summary_js.pb";

		// output protobuf message in binary format
		fs.writeFileSync(moduleResultPbFname, result.serializeBinary());
		fs.writeFileSync(moduleSummaryPbFname, summary.serializeBinary());
	}
}
